src/app/data-model-provider/EventsGenerator.h

```diff
@@ -46,60 +50,28 @@ class SimpleEventPayloadWriter : public EventLoggingDelegate
 std::optional<EventNumber> GenerateEvent(G & generator, const T & aEventData, EndpointId aEndpoint)
 {
     internal::SimpleEventPayloadWriter<T> eventPayloadWriter(aEventData);
-    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
     EventOptions eventOptions;
-    eventOptions.mPath        = path;
+    eventOptions.mPath        = ConcreteEventPath(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
     eventOptions.mPriority    = aEventData.GetPriorityLevel();
     eventOptions.mFabricIndex = aEventData.GetFabricIndex();
-
-    // this skips generating the event if it is fabric-scoped but the provided event data is not
-    // associated with any fabric.
-    if (eventOptions.mFabricIndex == kUndefinedFabricIndex)
-    {
-        ChipLogError(EventLogging, "Event encode failure: no fabric index for fabric scoped event");
-        return std::nullopt;
-    }
-
-    //
-    // Unlike attributes which have a different 'EncodeForRead' for fabric-scoped structs,
-    // fabric-sensitive events don't require that since the actual omission of the event in its entirety
-    // happens within the event management framework itself at the time of access.
-    //
-    // The 'mFabricIndex' field in the event options above is encoded out-of-band alongside the event payload
-    // and used to match against the accessing fabric.
-    //
-    EventNumber eventNumber;
-    CHIP_ERROR err = generator.GenerateEvent(&eventPayloadWriter, eventOptions, eventNumber);
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(EventLogging, "Failed to generate event: %" CHIP_ERROR_FORMAT, err.Format());
-        return std::nullopt;
-    }
-
-    return eventNumber;
+    return GenerateEvent(eventOptions, generator, eventPayloadWriter, true /* isScopedEvent */);
 }
 
 template <typename G, typename T, std::enable_if_t<!DataModel::IsFabricScoped<T>::value, bool> = true>
-std::optional<EventNumber> GenerateEvent(G & generator, const T & aEventData, EndpointId endpointId)
+std::optional<EventNumber> GenerateEvent(G & generator, const T & aEventData, EndpointId aEndpoint)
 {
     internal::SimpleEventPayloadWriter<T> eventPayloadWriter(aEventData);
-    ConcreteEventPath path(endpointId, aEventData.GetClusterId(), aEventData.GetEventId());
     EventOptions eventOptions;
-    eventOptions.mPath     = path;
+    eventOptions.mPath     = ConcreteEventPath(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
     eventOptions.mPriority = aEventData.GetPriorityLevel();
-    EventNumber eventNumber;
-    CHIP_ERROR err = generator.GenerateEvent(&eventPayloadWriter, eventOptions, eventNumber);
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(EventLogging, "Failed to generate event: %" CHIP_ERROR_FORMAT, err.Format());
-        return std::nullopt;
-    }
-
-    return eventNumber;
+    return GenerateEvent(eventOptions, generator, eventPayloadWriter, false /* isScopedEvent */);
```

Comment:
Can you try having isScopedEvent as part of eventOptions? Does it save more flash?

Reply:
Likely yes, but eventOptions is a shared existing class, so have to check blast radius here...